### PR TITLE
FS: enhance file_allocate_from_tcb() mulit-threads saftey & remove memcpy when close the fd

### PR DESCRIPTION
**Pull request by GUIDINGLI**
## Summary

FS: enhance file_allocate_from_tcb() mulit-threads saftey & remove memcpy when close the fd

## Impact

FS

## Testing

BES boards

### CODE REVIEW

File: fs/inode/fs_files.c [hunk 1]
```diff
@@ -375,64 +394,54 @@ int file_allocate_from_tcb(FAR struct tcb_s *tcb, FAR struct inode *inode,
...
-    {
       do
         {
-          filep = files_fget_by_index(list, i, j);
+          filep = &list->fl_files[i][j];
           if (filep->f_inode == NULL)
             {
+              filep->f_oflags = oflags;
+              filep->f_pos    = pos;
+              filep->f_inode  = inode;
```
Comment (anchao) [inline fs/inode/fs_files.c:434]: keep `filep->f_inode  = inode;` here and move f_oflags/f_pos/f_priv assign to line 445

**Review by anchao — APPROVED** [review]

**Review by acassis — APPROVED** [review]

**Comment by acassis** [comment]
@GUIDINGLI I liked it removed two nested while()s with a single for() and it seems to simplify and improve performance. I'm afraid of the impacts, did you test using ostest and more extensive testing?
Also since Tiago noticed the stack usage impact caused by #11777 could you confirm this modification will not increase code and/or stack usage too much?

**Comment by xiaoxiang781216** [comment]
@GUIDINGLI please rebase your patch.

**Comment by GUIDINGLI** [comment]
> @GUIDINGLI I liked it removed two nested while()s with a single for() and it seems to simplify and improve performance. I'm afraid of the impacts, did you test using ostest and more extensive testing? Also since Tiago noticed the stack usage impact caused by #11777 could you confirm this modification will not increase code and/or stack usage too much?

@acassis 
The original code also has two nested while()s, and I don't think it will take more stack usage then the previous.
And this PR's purpose is enhance the safety, because:
file_allocate_from_tcb() hasn't fully protected by spin_lock_irqsave().

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**